File: python/app/service.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Protocol

from .errors import InvalidVitalError
from .models import Alert, Event, EventType, Vital
from .store import Store


class Publisher(Protocol):
    def publish(self, event: Event) -> None:
        ...


class Service:
    def __init__(self, store: Store, publisher: Publisher) -> None:
        self._store = store
        self._publisher = publisher

    def ingest_vital(
        self,
        patient_id: str,
        systolic: int,
        diastolic: int,
        taken_at: datetime | None,
    ) -> Vital:
        patient_id = (patient_id or "").strip()
        if not patient_id:
            raise InvalidVitalError("invalid vital: patient_id is required")
        if systolic <= 0 or diastolic <= 0:
            raise InvalidVitalError("invalid vital: systolic and diastolic must be positive")
        if taken_at is None:
            raise InvalidVitalError("invalid vital: taken_at is required")

        if taken_at.tzinfo is None:
            taken_at = taken_at.replace(tzinfo=timezone.utc)
        taken_at = taken_at.astimezone(timezone.utc)

        vital = Vital(
            patient_id=patient_id,
            systolic=systolic,
            diastolic=diastolic,
            taken_at=taken_at,
            received_at=datetime.now(timezone.utc),
        )

        stored = self._store.add_vital(vital)
        event = Event(type=EventType.VITAL_RECEIVED, vital=stored)
        self._publisher.publish(event)
        return stored

    def list_alerts(self, patient_id: str) -> list[Alert]:
        alerts = self._store.list_alerts()
        patient_id = (patient_id or "").strip()
        if not patient_id:
            return alerts
        return [alert for alert in alerts if alert.patient_id == patient_id]

    def list_vitals(self, patient_id: str) -> list[Vital]:
        vitals = self._store.list_vitals()
        patient_id = (patient_id or "").strip()
        if not patient_id:
            return vitals
        return [vital for vital in vitals if vital.patient_id == patient_id]

File: node/src/domain/service.ts
import { InvalidVitalError, RequestError } from "./errors";
import {
  Alert,
  Event,
  EventTypeVitalReceived,
  Vital,
  VitalInput,
} from "./models";
import { ListOptions, ListResult, Store } from "./store";

export interface Publisher {
  publish(event: Event): Promise<void>;
}

export class Service {
  constructor(private store: Store, private publisher: Publisher) {}

  async ingestVital(
    patientId: string,
    systolic: number,
    diastolic: number,
    takenAt: Date,
    signal?: AbortSignal
  ): Promise<Vital> {
    const trimmed = (patientId || "").trim();
    if (!trimmed) {
      throw new InvalidVitalError("invalid vital: patientId is required");
    }
    if (systolic <= 0 || diastolic <= 0) {
      throw new InvalidVitalError(
        "invalid vital: systolic and diastolic must be positive"
      );
    }
    if (!takenAt || Number.isNaN(takenAt.getTime())) {
      throw new InvalidVitalError("invalid vital: takenAt is required");
    }

    const vital: VitalInput = {
      patientId: trimmed,
      systolic,
      diastolic,
      takenAt: new Date(takenAt.getTime()),
      receivedAt: new Date(),
    };

    const stored = await this.store.addVital(vital, signal);
    const event: Event = { type: EventTypeVitalReceived, vital: stored };
    await this.publisher.publish(event);
    return stored;
  }

  async listAlerts(
    options: ListOptions,
    signal?: AbortSignal
  ): Promise<ListResult<Alert>> {
    const patientId = (options.patientId || "").trim();
    if (!patientId) {
      throw new RequestError("patientId is required");
    }
    return this.store.listAlerts({ ...options, patientId }, signal);
  }

  async listVitals(
    options: ListOptions,
    signal?: AbortSignal
  ): Promise<ListResult<Vital>> {
    const patientId = (options.patientId || "").trim();
    if (!patientId) {
      throw new RequestError("patientId is required");
    }
    return this.store.listVitals({ ...options, patientId }, signal);
  }
}
